### ISSUE TITLE
Question : How to fix udp stream after CC errors

### ISSUE THREAD

**Issue by Inlarion** [tsduck/tsduck#560]
Hello! I apologize for the very poor English.
We receive via the optical udp multicast network from the television provider, he, in turn, receives these streams from the satellite.
Periodically, their satellite equipment receives interference, because of this, CC errors are present in the streams.
We get multicast and use ffmpeg to convert them to hls, then this hls is taken away by stb based on linux and android, and of course vlc.

Sometimes when CC errors occur, in the ffmpeg log we observe the following errors:
Non-monotonous DTS in output stream 0: 1 and / or Invalid timestamps
If after the appearance of these errors do not restart ffmpeg at the output we get the audio and video out of sync ...

The various ffmpeg options have no positive effect.

Linux-based STB, VLC show these streams without any problems (there are almost no problems with the image)
But STBs based on Android 6-7 with Stagefright cannot survive this, the picture hangs tight.

I would like to replace the heavy and uncomfortable ffmpeg with something more specialized, and if TSDuck can also fix these errors, then my happiness will simply have no limit.

**Comment by lars18th** [comment]
Hi @Inlarion ,

Yes, you can fix theses problems with TSDuck! :smile: 
I recommend to read the documentation about the multiple plugins. However, some tips:

- To fix CC errors (and only CC errors, this doesn't restore missing A/V packets):
`$ tsp -I ip 239.0.0.1:1234 -P continuity --fix -O ip 239.0.0.2.:1234`
 or if you want to ingest directly in FFmepg:
`$ tsp -I ip 239.0.0.1:1234 -P continuity --fix | ffmpeg ...`

- To generate the HLS stream directly from TSDuck:
`$ tsp -I ip 239.0.0.1:1234 -P continuity --fix -P zap 1 -O hls --live stream.hls`
However, in this case read the documentation about the multiple parameters of the HLS output.

Please, note that these commands are only _examples_. You need to adapt them to your specific environment.

Regards.

**Comment by barhom** [comment]
@lars18th What are the downside of using fixcc in tsduck?
I mean't it isn't like tsduck can "magically" fix these CC errors.

All it does is hide them?

**Comment by lelegard** [comment]
> I mean't it isn't like tsduck can "magically" fix these CC errors. All it does is hide them?

Yes, definitely, you are right. All commands and plugins in TSDuck are low-level tools which must be used by engineers who are fully aware of what they are doing. `tsfixcc` recomputes CC order to make them continuous. That's what it does, nothing more.

- If you have concatenated two valid TS with compatible structures and PID's and you want to erase the discontinuity in the middle, `tsfixcc` will do the job. 
- If you receive a stream with dropped packets, `tsfixcc` will certainly not recreate the missing packets 😄 

TSDuck provides the tools. You use them. The day TSDuck will come with a hammer, you can use it to drive a nail, smash a window or even kill you neighbour. Your choice...

**Comment by barhom** [comment]
>>> TSDuck provides the tools. You use them. The day TSDuck will come with a hammer, you can use it to drive a nail, smash a window or even kill you neighbour. Your choice...

I like your reasoning ;)

I really do appreciate the product. tsduck is awesome.

Can I ask you about what you know about IAT?
We use DVB-S tuners to downlink satellite channels that we then forward to operators over zixi or srt links. They all complain about their probes raising alarms about inter-arrival-time or IAT errors.

Is this because we downlink using commodity servers and pci-e tuners and not using expensive headends? Can tsduck help with IAT or at least measure it?

**Comment by lelegard** [comment]
> Is this because we downlink using commodity servers and pci-e tuners and not using expensive headends?

In my experience, commodity servers and PCIe tuners are fine and stable (much better that most USB tuners). I developed TSDuck precisely to avoid expensive equipment. And also to have a flexible framework which can be easily extended to provide the exact services I needed.

>  Can tsduck help with IAT or at least measure it?

You may try to use `tsp` to receive from plugin `dvb` and send through plugin `srt` (I know nothing about zixi). Adjust options `--max-input-packets` and `--max-flushed-packets` to sufficiently low values to smoothen the output.
